src/test/CropManager-unit.t.sol

```diff
@@ -0,0 +1,458 @@
+        (Usr a,) = init_user();
+        a.join(100 * 1e6);
+        a.frob(address(a), address(this), address(a), 100 * 1e18, 50 * 1e18);
+    }
+    function testFail_frob3() public {
+        (Usr a,) = init_user();
+        a.join(100 * 1e6);
+        a.frob(address(a), address(a), address(this), 100 * 1e18, 50 * 1e18);
+    }
+
+    function test_flux_other() public {
+        (Usr a, Usr b) = init_user();
+        a.join(100 * 1e6);
+        assertEq(a.gems(), 100 * 1e18);
+        assertEq(a.stake(), 100 * 1e18);
+        a.flux(address(a), address(b), 100 * 1e18);
+        assertEq(b.gems(), 100 * 1e18);
+        assertEq(b.stake(), 100 * 1e18);
+        b.exit(100 * 1e6);
+        assertEq(b.gems(), 0);
+        assertEq(b.stake(), 0);
+        assertEq(gem.balanceOf(address(b)), 300 * 1e6);
+    }
+
+    function test_flux_yourself() public {
+        // Flux to yourself should be a no-op
+        (Usr a,) = init_user();
+        a.join(100 * 1e6);
+        uint256 crops = a.crops();
+        assertEq(a.gems(), 100 * 1e18);
+        assertEq(a.stake(), 100 * 1e18);
+        a.flux(address(a), address(a), 100 * 1e18);
+        assertEq(a.gems(), 100 * 1e18);
+        assertEq(a.stake(), 100 * 1e18);
+        assertEq(a.crops(), crops);
+        a.exit(100 * 1e6);
+        assertEq(a.gems(), 0);
+        assertEq(a.stake(), 0);
+        assertEq(gem.balanceOf(address(a)), 200 * 1e6);
+    }
+
+    // Non-msg.sender srcs for flux should be disallowed for now
+    function testFail_flux() public {
+        (Usr a, Usr b) = init_user();
+        b.join(100 * 1e6);
+        a.flux(address(b), address(a), 100 * 1e18);
+    }
+
+    function testFail_quit() public {
+        (Usr a,) = init_user();
+        a.join(100 * 1e6);
+        a.frob(100 * 1e18, 50 * 1e18);
+        a.quit();       // Attempt to unbox the urn (should fail when vat is live)
+    }
+
+    function test_quit() public {
+        (Usr a,) = init_user();
+        a.join(100 * 1e6);
+        a.frob(100 * 1e18, 50 * 1e18);
+        vat.cage();
+        (uint256 ink, uint256 art) = vat.urns(ilk, a.proxy());
+        assertEq(ink, 100 * 1e18);
+        assertEq(art, 50 * 1e18);
+        assertEq(adapter.stake(a.proxy()), 100 * 1e18);
+        (ink, art) = vat.urns(ilk, address(a));
+        assertEq(ink, 0);
+        assertEq(art, 0);
+        assertEq(vat.gem(ilk, address(a)), 0);
+        assertEq(adapter.stake(address(a)), 0);
+        a.quit();
+        (ink, art) = vat.urns(ilk, a.proxy());
+        assertEq(ink, 0);
+        assertEq(art, 0);
+        assertEq(adapter.stake(a.proxy()), 100 * 1e18);
+        (ink, art) = vat.urns(ilk, address(a));
+        assertEq(ink, 100 * 1e18);
+        assertEq(art, 50 * 1e18);
+        assertEq(vat.gem(ilk, address(a)), 0);
+        assertEq(adapter.stake(address(a)), 0);
+        
+        // Can now interact directly with the vat to exit
+
+        a.frobDirect(address(a), address(a), address(a), -100 * 1e18, -50 * 1e18);
```

Comment:
This part of the test is a bit weird, as the real `vat.frob` won't work after shutdown.

Reply:
Its true that it represents something similar as what `end.skim` will do, so maybe is fine.